RICH TEXT EDITOR: Add a prosemirror to enable rich text editing

ajwl opened the pull request:
## What's changed?
A new rich text editor for the Headline input on some cards. 

* Only available on Email fronts 

* Once other work is completed, this should only be available on a new type of "free text card" (see #1097 ).

Email editors often add free text sections representing "notes from the editor", or some introductory text. They often want to add links or bold / italics in here. Doing it in the current section involves handwriting html which gets tedious. The new editor should make this easier and quicker. 

![Screenshot 2019-11-15 at 11 36 14](https://user-images.githubusercontent.com/10324129/68940853-2a0cdc80-079c-11ea-8139-f4c0ba9c4348.png)

## Implementation notes
<!-- Include any specific areas you want to highlight for review that you feel might be worthy of discussion (i.e. any non-obvious decisions you've made) -->
* This includes a lot of code lifted and shifted from https://github.com/guardian/flexible-content (specifically from the [prosemirror-setup](https://github.com/guardian/flexible-content/tree/master/composer/src/js/prosemirror-setup) directory) with minimal changes made, aside from to limit the types of styling that can be applied. 

* Elements that can form part of outputted HTML are `<strong>`,`<em>`, `<a>`, `<p>` and `&nbsp;`

* The final effect of the outputted html will be seen in the email renderer and should be checked. 

* Integration test added for bolding/unbolding and adding/removing a link 

## Checklist

### General
- [ ] 🤖 Relevant tests added
- [ ] ✅ CI checks / tests run locally
- [ ] 🔍 Checked on CODE

### Client
- [ ] 🚫 No obvious console errors on the client (i.e. React dev mode errors)
- [ ] 🎛️ No regressions with existing user interactions (i.e. all existing buttons, inputs etc. work)
- [ ] 📷 Screenshots / GIFs of relevant UI changes included

ajwl commented:
@jennygrahamjones so the CSS problem must be something to do with this..... BUT I DON'T KNOW WHAT 

`A custom plugin that adds a menuContent prop for the prosemirror-menu wrapper, and **a CSS class that enables the additional styling defined in style/style.css in this package**`

https://github.com/ProseMirror/prosemirror-example-setup

ajwl commented:
Passing over to @jennygrahamjones 🙏

ajwl commented on client-v2/integration/tests/main.spec.js:
```diff
@@ -76,16 +76,6 @@ test('Drag from feed to group position', async t => {
-    .expect(collectionLaunchButton(1).exists)
-    .eql(true);
-});
-
```
Noticed that this test was duplicated (see below) so just decided to remove it

ajwl commented on client-v2/src/components/FrontsEdit/CardFormInline.tsx:
```diff
@@ -404,7 +406,9 @@ class FormComponent extends React.Component<Props, FormComponentState> {
                 placeholder={articleCapiFieldValues.headline}
-                component={InputTextArea}
+                component={
+                  this.props.isEmailFronts ? RichTextInput : InputTextArea
```
currently the only check we make here is the for the email priority. Once type of snap is added this can go here too

jonathonherbert replied:
Yes, this is coming in #1105.

ajwl commented:
@jonathonherbert -> test added. Believe this is now ready for review. 
A lot of code copied and pasted from composer as per comment above... Let me know what you think on Monday

jonathonherbert commented on client-v2/src/components/inputs/RichTextInput.tsx:
```diff
@@ -0,0 +1,75 @@
+    if (!props.input.value) {
+      return;
+    }
+    // Editor view takes an HTML Node therefore this string value needs to be converted into a node by placing in a di
```
```suggestion
    // Editor view takes an HTML Node therefore this string value needs to be converted into a node by placing in a div
```

jonathonherbert replied:
nittiest of nits

jonathonherbert commented on client-v2/src/components/inputs/richtext/MenuView.tsx:
```diff
@@ -0,0 +1,100 @@
+    },
+    {
+      command: (state: EditorState, dispatch: (tr: Transaction) => void) =>
+        unlinkItemCommand(basicSchema.marks.link)(state, dispatch),
```
If I'm right, this could be simplified to:

```ts
command: unlinkItemCommand(basicSchema.marks.link)
```

ajwl replied:
Ah yeah it can 👍

jonathonherbert commented on client-v2/src/components/inputs/richtext/command-helpers.ts:
```diff
@@ -0,0 +1,149 @@
+import { MarkType } from 'prosemirror-model';
+import { EditorState, Transaction } from 'prosemirror-state';
+
+// These prosemirror-helper functions are a simplified version of what we use in Composer, and have been lifted and shifted from that repo
```
This is useful to know -- possibly we could think about a shared repo for these things, as PM ends up in more places we'll need these things again and again

ajwl commented:
@jonathonherbert Great - thanks a lot for the feedback. Will address this now

ajwl commented:
Code changes made. Changed margin on the menu, and added a message to the link validator: 
![Screenshot 2019-11-19 at 11 53 17](https://user-images.githubusercontent.com/10324129/69144482-4de56080-0ac3-11ea-8d48-db46ecd74597.png)

![Screenshot 2019-11-19 at 11 52 49](https://user-images.githubusercontent.com/10324129/69144497-550c6e80-0ac3-11ea-9e40-49231d428199.png)
